[ADD] inventory: add before date res doc (#7925)
odoo/documentation: content/applications/inventory_and_mrp/inventory/shipping_receiving/reservation_methods/before_scheduled_date.rst

=================================
Before scheduled date reservation
=================================

The *Before scheduled date* reservation method allows users to select a specific number of days that
will act as the maximum number of days **before** a scheduled delivery date that products included

> in a sales order (SO) should be reserved.

Author: Suggestion implemented ✅

.. seealso::
   :doc:`About reservation methods <../reservation_methods>`

Configuration
=============

To set the reservation method to *Before scheduled date*, navigate to :menuselection:`Inventory app
--> Configuration --> Operations Types`.

In the :guilabel:`General` tab, locate the :guilabel:`Reservation Method` field, and select

> :guilabel:`Before scheduled date`.

Author: Yes, I added in a note admonition explaining this:

.. note::
   When the :guilabel:`Type of Operation` is changed to *Receipt* on an :guilabel:`Operations Type`
   form, reservation methods are **not** available.
   
  The only other "restrictions" would be settings that need to be enabled for other operations types to appear, which is outlined in the Reservation Methods overview/parent doc (linked above this).

Suggestion implemented ✅

.. image:: before_scheduled_date/before-scheduled-date-configuration.png
   :align: center
   :alt: Reservation method field on delivery order operation type form.

Once selected, a new :guilabel:`Reserve before scheduled date` field appears below. From this field,
the number of :guilabel:`days before` and :guilabel:`days before when starred` can be changed from
the default `0`.

Changing the :guilabel:`days before` value changes the maximum number of days before a scheduled
date that products should be reserved.

> Changing the :guilabel:`days before when starred` value changes the maximum number of days before
> a scheduled date that starred (favorited) transfers for products should be reserved.

Reviewer: ```suggestion
Changing the :guilabel:`days before when starred` value changes the maximum number of days before
a scheduled date that products should be reserved for starred (favorited) transfers.
```

Author: Suggestion implemented ✅

.. example::
   In the image below, the :guilabel:`days before` value is set to `2` days before, and the
   :guilabel:`days before when starred` value is set to `3`.

   This means that products will be reserved two days before the scheduled delivery date for normal

>    orders, and three days before the scheduled delivery date for starred (favorited) transfers.

Author: I didn't know that before, but now I do 👍 Suggestion implemented ✅

   .. image:: before_scheduled_date/before-scheduled-date- days-before.png
      :align: center
      :alt: Reserve before scheduled date field with set numerical values.

   This is the configuration applied for the following workflow below.

Edit product form
-----------------

Before beginning the workflow, first add a *customer lead time* to the product that will be sold. To

> do so, navigate to :menuselection:`Products --> Products`, and select the desired product.

Author: Yeah, that's what I meant to convey here, so the clarifying language is helpful. Suggestion implemented ✅

Click the :guilabel:`Inventory` tab, and under the :guilabel:`Logistics` section, change the value
in the :guilabel:`Customer Lead Time` field to `5` days. This sets the scheduled delivery date for

> this specific product to five days after the creation date of the sales order.

Author: I understand that logic, it does make sense in this instance. Suggestion implemented ✅

.. image:: before_scheduled_date/before-scheduled-date-customer-lead-time.png
   :align: center
   :alt: Product form with customer lead time set in Inventory tab.

Workflow
========

To see the *Before scheduled date* reservation method in action, create a new :abbr:`SO (sales
order)` by navigating to :menuselection:`Sales app --> New`.

Add a customer in the :guilabel:`Customer` field, then, in the :guilabel:`Order Lines` tab, click
:guilabel:`Add a product`, and select a product to add to the quotation from the drop-down menu.

> Finally, in the :guilabel:`Quantity` column, adjust the desired quantity of the product to sell.

Author: Sounds good to me. Suggestion implemented ✅

Once ready, click :guilabel:`Confirm` to confirm the sales order.

> Click the green :guilabel:`forecast graph` icon on the product line to reveal the product's

Reviewer: ```suggestion
Click the green :guilabel:` 📈 (forecast graph)` icon on the product line to reveal the product's
```

Author: Suggestion implemented ✅

:guilabel:`Availability` tooltip. This tooltip reveals the reserved number of units for this order.
Because the reservation method is set to *Before scheduled date*, the :guilabel:`Reserved` quantity
reads `0 Units`.

However, below that quantity reads `Available in stock`. This is because the quantity is available,
but the scheduled date is five days from the order date. Since reservation isn't until two days

> before the scheduled delivery, it will not reserve the products until then.

Author: Yes, that is definitely helpful to clarify - I added in the following admonition to explain:

.. note::
   If there is **not** sufficient quantity of stock for the product included in the |SO|, the
   :guilabel:`📈 (forecast graph)` icon is red, instead of green.

   Instead of revealing the reserved number of units for the order, the :guilabel:`Availability`
   tooltip reads :guilabel:`Reserved`, and reveals the available number of units (e.g., `0 Units`).

   Additionally, unless there is a set replenishment or a live receipt, it also reads :guilabel:`No
   future availability`, in red text.

Suggestion implemented ✅

.. image:: before_scheduled_date/before-scheduled-date-availability-tooltip.png
   :align: center
   :alt: Confirmed sales order with product availability tooltip selected.

Click the :guilabel:`Delivery` smart button to see the delivery order form.

On the :guilabel:`Delivery Order` form, the status in the :guilabel:`Product Availability` field is
listed as `Available`, in yellow text instead of green. This is because there is sufficient stock on

> hand for this order, but no quantity has been reserved yet.

Reviewer 2: removed the GUI from 'Delivery Order' because that isn't shown/displayed on the screen/UI

> hand for this order, but no quantity has been reserved yet.

Author: Suggestion implemented ✅

Note the :guilabel:`Scheduled Date` field above the :guilabel:`Product Availability` field displays
the date five days from the order creation date. This indicates that the products will not be

> reserved until three days from today's date (two days before the scheduled delivery date).

Author: Suggestion implemented ✅

.. image:: before_scheduled_date/before-scheduled-date-delivery-order-form.png
   :align: center
   :alt: Delivery order form with product availability and reserved quantity.

In the :guilabel:`Operations` tab on the :guilabel:`Product` line, the numbers in the

> :guilabel:`Demand` column and the :guilabel:`Quantity` column do no match (in this case, the

Reviewer: ```suggestion
:guilabel:`Demand` column and the :guilabel:`Quantity` column do not match (in this case, the
```

Author: Suggestion implemented ✅

> :guilabel:`Demand` column lists `10.00`, while the :guilabel:`Quantity` column lists `0`.

Reviewer 2: Considering this is where the sample flow ends currently, it seems a bit "unfinished." Perhaps you could add a mention of how, when the time comes, the numbers in the columns will match...? Or something along those lines. It just seems like a bit of "conclusion" is needed to round things out a touch.

.. tip::
   If the product(s) in the :abbr:`SO (sales order)` should be reserved sooner than the scheduled
   reservation date, the reservation can be manually "overridden". To manually reserve the products

>    sooner than scheduled, click :guilabel:`Check Availability` at the top of the form.

Author: Yeah, I just thought it would be useful to indicate that it can apply to both orders with 1 product, and orders with multiple. But I'll stick to the rules here, Suggestion Implemented ✅

   This turns the `Available` status in the :guilabel:`Product Availability` field green, and

>    changes the number in the :guilabel:`Quantity` column to `10.00`, so it matches the
>    :guilabel:`Demand` column.

Reviewer 2: ```suggestion
   changes the number in the :guilabel:`Quantity` column to match the :guilabel:`Demand` column.
```

Author: Suggestion implemented ✅

=================================
Before scheduled date reservation
=================================

.. _inventory/reservation_methods/before-scheduled-date:

.. |SO| replace:: :abbr:`SO (Sales Order)`

The *Before scheduled date* reservation method allows users to select a specific number of days that
act as the maximum number of days **before** a scheduled delivery date, when products included in a
sales order (SO) should be reserved.

.. seealso::
   :doc:`About reservation methods <../reservation_methods>`

Configuration
=============

To set the reservation method to *Before scheduled date*, navigate to :menuselection:`Inventory app
--> Configuration --> Operations Types`. Then, select the desired :guilabel:`Operation Type` to
configure, or create a new one by clicking :guilabel:`New`.

In the :guilabel:`General` tab, locate the :guilabel:`Reservation Method` field, and select
:guilabel:`Before scheduled date`.

.. image:: before_scheduled_date/before-scheduled-date-configuration.png
   :align: center
   :alt: Reservation method field on delivery order operation type form.

.. note::

>    When the :guilabel:`Type of Operation` is changed to *Receipt* on an :guilabel:`Operations Type`
>    form, reservation methods are **not** available.

Reviewer 2: ```suggestion
   When the :guilabel:`Type of Operation` is changed to :guilabel:`Receipt` on an
   :guilabel:`Operations Type` form, reservation methods are **not** available.
```

Author: Suggestion implemented ✅

Once selected, a new :guilabel:`Reserve before scheduled date` field appears below. From this field,
the number of :guilabel:`days before` and :guilabel:`days before when starred` can be changed from
the default `0`.

Changing the :guilabel:`days before` value changes the maximum number of days before a scheduled
date that products should be reserved.

Changing the :guilabel:`days before when starred` value changes the maximum number of days before

> a scheduled date that products should be reserved for starred (favorited) transfers.

Reviewer 2: ```suggestion
a scheduled date that products should be reserved if the transfers are starred (favorited).
```

Author: Suggestion implemented ✅

> Changing the :guilabel:`days before when starred` value changes the maximum number of days before
> a scheduled date that products should be reserved if the transfers are starred (favorited).

Reviewer 3: Early break :)
```suggestion
Changing the :guilabel:`days before when starred` value changes the maximum number of days before a
scheduled date that products should be reserved if the transfers are starred (favorited).
```

Author: Suggestion implemented ✅

.. example::

>    Here, the :guilabel:`days before` value is set to `2` days before, and the :guilabel:`days before
>    when starred` value is set to `3`.

Reviewer 2: ```suggestion
   If the :guilabel:`days before` value is set to `2` days before, and the :guilabel:`days before
   when starred` value is set to `3`.
```
I know you implemented my previous suggestion, but I think this current one is a bit better -- your call.

Author: Suggestion implemented ✅

   This means products are reserved two days before the scheduled delivery date for normal orders,
   and three days before the scheduled delivery date for starred (favorited) transfers.

   .. image:: before_scheduled_date/before-scheduled-date- days-before.png
      :align: center
      :alt: Reserve before scheduled date field with set numerical values.

>    This is the configuration applied for the following workflow below.

Reviewer 2: ```suggestion
   This is the configuration applied for the following workflow found below.
```

Edit product form
-----------------

Before the *Before scheduled date* reservation method can be used, ensure that a *customer lead
time* is added to products that plan to be sold with this method.

To do that, navigate to :menuselection:`Inventory app --> Products --> Products`, and select the
desired product to configure.

On the product form, click the :guilabel:`Inventory` tab, and under the :guilabel:`Logistics`
section, change the value in the :guilabel:`Customer Lead Time` field.

For this example workflow, change it to `5` days.

This sets the scheduled delivery date for this specific product to five days after the creation date
of the sales order.

.. image:: before_scheduled_date/before-scheduled-date-customer-lead-time.png
   :align: center
   :alt: Product form with customer lead time set in Inventory tab.

Workflow
========

To see the *Before scheduled date* reservation method in action, create a new |SO| by navigating to
:menuselection:`Sales app --> New`.

Add a customer in the :guilabel:`Customer` field, then, in the :guilabel:`Order Lines` tab, click

> :guilabel:`Add a product`, and select a product, with a configured *customer lead time*, to add to
> the quotation from the drop-down menu.

Reviewer 2: ```suggestion
:guilabel:`Add a product`, and select a product from the drop-down menu that has a configured
*customer lead time*, to add to the quotation form.
```

Author: Suggestion implemented ✅

Finally, in the :guilabel:`Quantity` column, adjust the desired quantity of the product to sell.

For this sample workflow, set the :guilabel:`Quantity` to `10`.

Once ready, click :guilabel:`Confirm` to confirm the sales order.

Click the green :guilabel:`📈 (forecast graph)` icon on the product line to reveal the product's
:guilabel:`Availability` tooltip. This tooltip reveals the reserved number of units for this order.
Because the reservation method is set to *Before scheduled date*, the :guilabel:`Reserved` quantity
reads `0 Units`.

However, below that quantity reads `Available in stock`. This is because the quantity is available,
but the scheduled date, for this example workflow, is five days from the order date.

> Since reservation isn't until two days before the scheduled delivery, it will not reserve the
> products until then.

Reviewer 2: ```suggestion
Since reservation is not until two days before the scheduled delivery, it will not reserve the
products until then.
```

Author: Suggestion implemented ✅

   future availability`, in red text.

.. image:: before_scheduled_date/before-scheduled-date-availability-tooltip.png
   :align: center
   :alt: Confirmed sales order with product availability tooltip selected.

Click the :guilabel:`Delivery` smart button to see the delivery order form.

On the delivery order form, the status in the :guilabel:`Product Availability` field is listed as
`Available`, in yellow text, instead of green. This is because there is sufficient stock on-hand for
this order, but no quantity has been reserved yet.

Note the :guilabel:`Scheduled Date` field, above the :guilabel:`Product Availability` field,
displays the date five days from the order creation date. This indicates that the products are not
reserved until three days from today's date (two days before the scheduled delivery date).

.. image:: before_scheduled_date/before-scheduled-date-delivery-order-form.png
   :align: center
   :alt: Delivery order form with product availability and reserved quantity.

In the :guilabel:`Operations` tab on the :guilabel:`Product` line, the numbers in the
:guilabel:`Demand` column and the :guilabel:`Quantity` column do not match (in this case, the

> :guilabel:`Demand` column lists `10.00`, while the :guilabel:`Quantity` column lists `0`.

Reviewer 2: ```suggestion
:guilabel:`Demand` column lists `10.00`, while the :guilabel:`Quantity` column lists `0`).
```

Author: Suggestion implemented ✅

The :guilabel:`Quantity` column lists `0` because the products aren't reserved until two days
*before* their delivery date. Odoo automatically reserves the products once the scheduled date
arrives, at which point the :guilabel:`Demand` and :guilabel:`Quantity` columns will match.

.. tip::

>    If the products in the |SO| should be reserved sooner than the scheduled reservation date, the

Reviewer 2: ```suggestion
   If the products in the |SO| should be reserved *sooner* than the scheduled reservation date, the
```

>    reservation can be manually "overridden." To manually reserve the products sooner than scheduled,

Reviewer 2: ```suggestion
   reservation can be manually overridden. To manually reserve the products sooner than scheduled,
```

Author: Suggestion implemented ✅

   click :guilabel:`Check Availability` at the top of the form.

   This turns the `Available` status in the :guilabel:`Product Availability` field green, and
   changes the number in the :guilabel:`Quantity` column to match the :guilabel:`Demand` column.

   Once ready, click :guilabel:`Validate`.

.. seealso::

>    - :doc:`Manual reservation <../reservation_methods/manually>`
>    - :doc:`At confirmation reservation <../reservation_methods/at_confirmation>`

Reviewer 3: No need to back out of this directory and move right back into it 😄, if I suggested this in the past I apologize, the path can simply be:
```suggestion
   - :doc:`Manual reservation <manually>`
   - :doc:`At confirmation reservation <at_confirmation>`
```

Author: Suggestion implemented ✅
